REL-847441 - Add Relativity Forms Event Handler Examples

justinyh opened the pull request:
This PR adds Relativity Forms Event Handler examples to the repo. It includes a clear README with basic info, a table of contents, and a note stating that these are intended to be used for reference and/or templates and that they are not guaranteed to function out-of-the-box.

justinyh committed:
Add code examples to repo

justinyh committed:
Update README

justinyh committed:
Update guids and namespaces to be unique

mikedewoskin commented on examples/CONTRIBUTING.md:
```diff
@@ -0,0 +1,34 @@
+- Test the example thoroughly to ensure it works as expected.
+- Update the examples level README.md to add your new addition to the Contents section.
+
+## best Practices
```
"B"est Practices

justinyh replied:
Fixed!

cmardyla commented on examples/Console_AddButton/README.md:
```diff
@@ -0,0 +1,8 @@
+This event handler adds a button to a console programmatically. The button's value changes based on the value of a field, once saved.
+
+## Compatibility
+R1, likely Server 2022 & Server 2023.
```
I don't want to blow up the scope here but wondering if we should do some quick testing to definitively confirm whether these event handlers are compatible with Server.

justinyh replied:
I think it's a fair question. I mainly went back and forth between whether it was worth it for example code. If it is, I would probably need some help parallelizing this!

justinyh replied:
I removed the compatibility denotation. I reclarified that the goal of these examples is to not provide functionality but examples to reference and/or use as templates.

cmardyla commented on examples/Multi_FieldManipulation/src/multi-field-manipulation.js:
```diff
@@ -0,0 +1,30 @@
+		layoutData.splice(3, 1); //assuming the category you want to remove is at the bottome of the layout, you can remove it like this.
+	};
+
+	//https://platform.relativity.com/RelativityOne/#Relativity_Forms/Load_pipeline.htm#hydrateL
```
This link is broken

cmardyla commented on examples/Multi_FieldManipulation/src/multi-field-manipulation.js:
The Platform doc links in this file are broken

justinyh replied:
Fixed them! Thanks for catching that!

cmardyla commented on examples/ActionBar_Add_Back_to_List_Button/RA_Relativity_Forms_Back_To_List_Button_20240208160809.rap:
Any reason why a RAP is included in this example but not the others? Should we be consistent?

justinyh replied:
@john-aldana was nice and included a RAP with the event handler he wrote 🥇 I think it was because this was provided to a customer to use.

It's quite a bit of manual work to create RAPs for each of these and I went back and forth as to whether it was worth it. Thoughts? If we do want them, I would also love to parallelize it a bit 😅

cmardyla replied:
Actually, if we end up testing this more extensively in Server, we'll probably end up creating RAPs anyway, so I suppose we might as well include them.

justinyh replied:
I decided against compatibility testing and reclarified the goals of these examples.

justinyh committed:
Fix typo in CONTRIBUTING.md and fix links in multi-field-manipulation.js

justinyh committed:
Consolidate .NET PIEH examples into one w/ multiple JS files, clarify contribution requirements

justinyh commented:
I have updated my PR to only have one template .NET Page Interaction Event Handler w/ multiple JS file examples to improve simplicity. I also have clarified that these are intended to be used as examples and/or templates and not for their intended functionality.

I changed my mind about the RAPs and the compatibility testing. The intent of these examples is not to provide functionality but to provide some examples to achieve common workflows. I think really this just needs a brief and basic review to ensure clarity exists and the code samples appear generally bug-free.

justinyh committed:
Remove compatibility info from back to list comment

mikedewoskin commented on README.md:
```diff
@@ -7,12 +7,15 @@ The Relativity Forms Development Tools repository's purpose is to house tools to
+* [Source](./Source/)
 
 ## Futures
 In the future, this repository will include a JavaScript library for unit testing Relativity Forms in your own development environments, and potentially also a project template for automation of your Relativity Forms development with NodeJS.
```
I know this is old text but would it make sense to be more noncommittal and say something like: In the future, this repository will be enhanced with additional examples and tools. Feedback from the community on what you would like to see here is appreciated.

mikedewoskin commented on examples/JsEventHandlerExamples/multi-field-manipulation.js:
```diff
@@ -0,0 +1,34 @@
+    };
+
+	eventHandlers[eventNames.PAGE_INTERACTION] = function() { 
+		console.log( "Inside PAGE_INTERACTION event handler" ); 
```
should we include a commented link to the pageInteraction documentation like the other events?

mikedewoskin commented on examples/README.md:
```diff
@@ -0,0 +1,39 @@
+# Relativity Forms Event Handler Examples
+This directory includes a set of example Relativity Forms Event Handlers.
+
+Note: These event handlers are meant to serve as samples and not all of them may work out of the box. Some may require setup. We recommend you use them as a template and/or reference, not for their intended functionality.
```
maybe bold Note

cmardyla commented on examples/JsEventHandlerExamples/actionBar-back-to-list.js:
```diff
@@ -0,0 +1,130 @@
+    * @returns {String|undefined} the URL to redirect to
+    */
+    function generateBackListPageUrl() {
+        const currentUrl = new URL(window.location.href);
```
Is it worth documenting that there may be more preferred ways of getting the URL in R1, and link to some PageBase docs?

cmardyla commented on examples/JsEventHandlerExamples/multi-field-manipulation.js:
```diff
@@ -0,0 +1,34 @@
+
+	// https://platform.relativity.com/RelativityOne/Content/FormsAPI/Load_Pipeline.htm#hydratelayout
+	eventHandlers[eventNames.HYDRATE_LAYOUT] = function(layoutData, objectInstanceData) {
+        console.log("Inside HYDRATE_LAYOUT event handler");
```
Fix indentation

cmardyla commented on examples/JsEventHandlerExamples/multi-field-manipulation.js:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Name: Field Manipulation Event Handlers
+ * Description: This is a set of event handlers that demonstrate the various methods to manipulate fields.
```
I don't remember the context around why this example was written but it feels like it could use more background info or some example scenarios for when this would be useful.

justinyh replied:
Tried my best with this. Let me know if you think there are improvements I can make here.

cmardyla replied:
Looks good to me, thanks!

cmardyla commented on examples/JsEventHandlerExamples/picker-override-filters.js:
```diff
@@ -0,0 +1,30 @@
+					};
+
+					return defaultGetDataFunction(request)
+						.then(function (response) {
```
Is this `.then` actually needed given that we're just returning the `response`?

cmardyla commented on examples/JsEventHandlerExamples/transformLayout-addField.js:
```diff
@@ -0,0 +1,92 @@
+				// should be useful for adding fields to json
+				const categories = convenienceApi.layout.getCategories(layoutData);
+
+				// when you add a field, are you adding it to the top of a category, the bottom of a category,
```
Could this comment be moved to the top of the file to spell out the prerequisites for this event  handler? Looks like we're assuming the layout has a category named "TR Text Two"

cmardyla commented on examples/JsEventHandlerExamples/transformLayout-addField.js:
```diff
@@ -0,0 +1,92 @@
+ * Description: This event handler adds a field to the layout programmatically.
+ * Note: This may need to be paired with a replaceSave event handler to ensure that the field is saved.
+ */
+(function(convenienceApi, EVENT_NAMES, privilegedEnvelope = {}) {
```
This event handler has a `priviledgedEnvelope` but the others don't - remove?

cmardyla commented on examples/PageInteractionEventHandler.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace FormsExamplePageInteractionEventHandler
+{
+	[kCura.EventHandler.CustomAttributes.Description("Relativity Forms Event Handler - This event handler adds a 'Back To List' button in Forms to attached applications")]
```
Update description attribute

justinyh committed:
PR feedback, syntax fixes, adding clarity

john-aldana commented on examples/JsEventHandlerExamples/transformLayout-addField.js:
```diff
@@ -0,0 +1,92 @@
+ * Description/Scenario: I'd like to programmatically add a field to my layout. I would like this field to go after the "TR Text Two" field in the same category.
+ * Note: This may need to be paired with a replaceSave event handler to ensure that the field is saved.
+ */
+(function(convenienceApi, EVENT_NAMES) {
```
this won't work

justinyh committed:
Fix eventNames/convenienceApi params in transformLayout-addField.js
